package io.github.finoid.snapshots;

import io.github.finoid.snapshots.config.BaseSnapshotConfig;
import io.github.finoid.snapshots.config.SnapshotConfig;
import io.github.finoid.snapshots.serializers.ToStringSnapshotSerializer;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.TestInfo;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Paths;

import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.api.Assertions.assertTrue;

public class OnLoadSnapshotFileTest {

    private static final String SNAPSHOT_FILE_PATH =
        "src/test/java/io/github/finoid/snapshots/__snapshots__/OnLoadSnapshotFileTest.snap";

    private static final SnapshotConfig CUSTOM_SNAPSHOT_CONFIG = new BaseSnapshotConfig();

    @BeforeAll
    static void beforeAll() throws IOException {
        Files.deleteIfExists(Paths.get(SNAPSHOT_FILE_PATH));
        String snapshotFileContent =
            "io.github.finoid.snapshots.OnLoadSnapshotFileTest.shouldLoadFileWithCorrectEncodingForCompare=[\n"
                + "any special characters that need correct encoding äöüèéàè\n"
                + "]";
        createSnapshotFile(snapshotFileContent);
    }

    private static void createSnapshotFile(String snapshot) {
        try {
            File file = new File(SNAPSHOT_FILE_PATH);
            file.getParentFile().mkdirs();
            file.createNewFile();
            try (FileOutputStream fileStream = new FileOutputStream(file, false)) {
                fileStream.write(snapshot.getBytes(StandardCharsets.UTF_8));
            } catch (IOException e) {
                throw new RuntimeException("Unable to write debug file ", e);
            }
        } catch (IOException e) {
            throw new RuntimeException("Unable to write debug file ", e);
        }
    }

    @DisplayName("Should load snapshots with correct encoding")
    @Test
    public void shouldLoadFileWithCorrectEncodingForCompare(TestInfo testInfo) throws IOException {
        assertTrue(Files.exists(Paths.get(SNAPSHOT_FILE_PATH)));

        SnapshotVerifier snapshotVerifier =
            new SnapshotVerifier(CUSTOM_SNAPSHOT_CONFIG, testInfo.getTestClass().get());
        Expect expect = Expect.of(snapshotVerifier, testInfo.getTestMethod().get());
        expect
            .serializer(ToStringSnapshotSerializer.class)
            .toMatchSnapshot("any special characters that need correct encoding äöüèéàè");
        snapshotVerifier.validateSnapshots();

        File f = new File(SNAPSHOT_FILE_PATH);
        assertThat(String.join("\n", Files.readAllLines(f.toPath())))
            .isEqualTo(
                "io.github.finoid.snapshots.OnLoadSnapshotFileTest.shouldLoadFileWithCorrectEncodingForCompare=[\n"
                    + "any special characters that need correct encoding äöüèéàè\n"
                    + "]");
    }
}
